packages/orderbook/src/order_provider/mesh_order_provider.ts

```diff
@@ -0,0 +1,207 @@
+     * assumed to be 0.
+     * @param orderEvent The `OrderEvent` from a Mesh subscription update
+     */
+    private static _orderInfoToAPIOrder(
+        orderEvent: OrderEvent | AcceptedOrderInfo | RejectedOrderInfo | OrderInfo,
+    ): APIOrder {
+        const remainingFillableTakerAssetAmount = (orderEvent as OrderEvent).fillableTakerAssetAmount
+            ? (orderEvent as OrderEvent).fillableTakerAssetAmount
+            : new BigNumber(0);
+        return {
+            order: orderEvent.signedOrder,
+            metaData: {
+                orderHash: orderEvent.orderHash,
+                remainingFillableTakerAssetAmount,
+            },
+        };
+    }
+
+    /**
+     * Instantiates a [Mesh](https://github.com/0xProject/0x-mesh) Order Provider. This provider writes
+     * all orders stored in Mesh to the OrderStore and subscribes all Mesh updates.
+     * @param opts `MeshOrderProviderOpts` containing the websocketEndpoint and additional Mesh options
+     * @param orderStore The `OrderStore` where orders are added and removed from
+     */
+    constructor(opts: MeshOrderProviderOpts, orderStore: OrderStore) {
+        super(orderStore);
+        this._wsClient = new WSClient(opts.websocketEndpoint, opts.wsOpts);
+    }
+
+    /**
+     * Returns the available asset pairs. If no subscription to Mesh exists (and therefor no orders) it is
+     * created and awaited on. Once the connection has been initialized the orders in the store are returned
+     * as asset pairs.
+     */
+    public async getAvailableAssetDatasAsync(): Promise<AssetPairsItem[]> {
+        await this._initializeIfRequiredAsync();
+        const assetPairsItems: AssetPairsItem[] = [];
+        const minAmount = new BigNumber(0);
+        const maxAmount = new BigNumber(2).pow(256).minus(1);
+        const precision = 18;
+        for (const assetPairKey of this._orderStore.keys()) {
+            const [assetA, assetB] = OrderStore.assetPairKeyToAssets(assetPairKey);
+            const assetDataA: Asset = { assetData: assetA, minAmount, maxAmount, precision };
+            const assetDataB: Asset = { assetData: assetB, minAmount, maxAmount, precision };
+            assetPairsItems.push({ assetDataA, assetDataB });
+            assetPairsItems.push({ assetDataA: assetDataB, assetDataB: assetDataA });
+        }
+        return assetPairsItems;
+    }
+
+    /**
+     * Retrieves the entire orderbook for the asset pair. If websocket is available
+     * a subscription will be created to keep these orders up to date. If no websocket
+     * connection is available calling this function will fetch orders from the SRA endpoint.
+     * @param makerAssetData the Maker Asset Data
+     * @param takerAssetData the Taker Asset Data
+     */
+    public async createSubscriptionForAssetPairAsync(_makerAssetData: string, _takerAssetData: string): Promise<void> {
+        // Create the subscription first to get any updates during waiting for the request
```

Comment:
```suggestion
        // Create the subscription first to get any updates while waiting for the request
```

This comment is confusing.